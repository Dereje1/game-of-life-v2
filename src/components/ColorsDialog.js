import React from "react";
import PropTypes from "prop-types";
import { SketchPicker } from "react-color";
import Button from "@mui/material/Button";
import DialogTitle from "@mui/material/DialogTitle";
import DialogContent from "@mui/material/DialogContent";
import DialogActions from "@mui/material/DialogActions";
import Dialog from "@mui/material/Dialog";
import ButtonGroup from "@mui/material/ButtonGroup";
import Typography from "@mui/material/Typography";
import IconButton from "@mui/material/IconButton";
import CancelOutlinedIcon from "@mui/icons-material/CancelOutlined";

const colorTypeMap = {
  canvasBackGround: "canvas",
  liveCell: "cell",
  grid: "grid"
};

const ColorsDialog = ({
  color,
  handleColorChange,
  closeColorPicker,
  open,
  updateColorChangeType,
  selectedColorType,
  showGrid
}) => (
  <Dialog
    sx={{ "& .MuiDialog-paper": { width: "auto", maxHeight: 530 } }}
    maxWidth="xs"
    open={open}
    onClose={closeColorPicker}
  >
    <DialogTitle sx={{ display: "flex", justifyContent: "center" }}>
      <Typography>{`Choose ${colorTypeMap[selectedColorType]} color`}</Typography>
    </DialogTitle>
    <DialogContent dividers sx={{ display: "flex", justifyContent: "center" }}>
      <SketchPicker color={color} onChange={handleColorChange} disableAlpha />
    </DialogContent>
    <DialogActions sx={{ justifyContent: "center" }}>
      <ButtonGroup variant="text" aria-label="outlined primary button group" fullWidth>
        {Object.keys(colorTypeMap).map((type) => (
          <Button
            onClick={() => updateColorChangeType(type)}
            color={selectedColorType === type ? "primary" : "inherit"}
            id={`${colorTypeMap[type]}-color-change`}
            key={type}
            disabled={type === "grid" && !showGrid ? true : false}
          >
            {colorTypeMap[type]}
          </Button>
        ))}
        <IconButton aria-label="reset" onClick={closeColorPicker}>
          <CancelOutlinedIcon style={{ fontSize: "2rem" }} />
        </IconButton>
      </ButtonGroup>
    </DialogActions>
  </Dialog>
);

ColorsDialog.propTypes = {
  color: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
  showGrid: PropTypes.bool.isRequired,
  selectedColorType: PropTypes.string.isRequired,
  handleColorChange: PropTypes.func.isRequired,
  closeColorPicker: PropTypes.func.isRequired,
  updateColorChangeType: PropTypes.func.isRequired
};

export default ColorsDialog;
