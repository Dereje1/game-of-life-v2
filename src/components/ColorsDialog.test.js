import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import ColorsDialog from "./ColorsDialog";

let props;
beforeEach(() => {
  props = {
    color: "blue",
    handleColorChange: jest.fn(),
    closeColorPicker: jest.fn(),
    open: true,
    updateColorChangeType: jest.fn(),
    selectedColorType: "canvasBackground",
    showGrid: true
  };
});

afterEach(() => {
  props = null;
  jest.clearAllMocks();
});

test("will render the settings dialog", () => {
  const wrapper = shallow(<ColorsDialog {...props} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test("will update the colorType to change for the canvas", () => {
  const wrapper = shallow(<ColorsDialog {...props} />);
  const canvasChange = wrapper.find({ id: "canvas-color-change" });
  canvasChange.props().onClick();
  expect(props.updateColorChangeType).toHaveBeenCalledWith("canvasBackground");
  expect(canvasChange.props().disabled).toBe(false);
});

test("will update the colorType to change for the live cells", () => {
  const wrapper = shallow(<ColorsDialog {...props} />);
  const cellChange = wrapper.find({ id: "cell-color-change" });
  cellChange.props().onClick();
  expect(props.updateColorChangeType).toHaveBeenCalledWith("liveCellColor");
  expect(cellChange.props().disabled).toBe(false);
});

test("will update the colorType to change for the grid", () => {
  const wrapper = shallow(<ColorsDialog {...props} />);
  const gridChange = wrapper.find({ id: "grid-color-change" });
  gridChange.props().onClick();
  expect(props.updateColorChangeType).toHaveBeenCalledWith("gridColor");
  expect(gridChange.props().disabled).toBe(false);
});

test("will disable grid colorType if grid is not showing", () => {
  const updatedProps = {
    ...props,
    showGrid: false
  };
  const wrapper = shallow(<ColorsDialog {...updatedProps} />);
  const gridChange = wrapper.find({ id: "grid-color-change" });
  expect(gridChange.props().disabled).toBe(true);
});
