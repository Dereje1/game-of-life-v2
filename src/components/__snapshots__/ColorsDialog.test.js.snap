// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render the settings dialog 1`] = `
<ForwardRef(Dialog)
  maxWidth="xs"
  onClose={[MockFunction]}
  open={true}
  sx={
    Object {
      "& .MuiDialog-paper": Object {
        "maxHeight": 530,
        "width": "auto",
      },
    }
  }
>
  <ForwardRef(DialogTitle)
    sx={
      Object {
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <ForwardRef(Typography)
      variant="subtitle"
    >
      Choose canvas color
    </ForwardRef(Typography)>
  </ForwardRef(DialogTitle)>
  <ForwardRef(DialogContent)
    dividers={true}
    sx={
      Object {
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <ColorPicker
      color="blue"
      disableAlpha={true}
      onChange={[MockFunction]}
      presetColors={
        Array [
          "#D0021B",
          "#F5A623",
          "#F8E71C",
          "#8B572A",
          "#7ED321",
          "#417505",
          "#BD10E0",
          "#9013FE",
          "#4A90E2",
          "#50E3C2",
          "#B8E986",
          "#000000",
          "#4A4A4A",
          "#9B9B9B",
          "#FFFFFF",
        ]
      }
      styles={Object {}}
      width={200}
    />
  </ForwardRef(DialogContent)>
  <ForwardRef(DialogActions)
    sx={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <ForwardRef(ButtonGroup)
      aria-label="outlined primary button group"
      fullWidth={true}
      variant="text"
    >
      <ForwardRef(Button)
        color="primary"
        disabled={false}
        id="canvas-color-change"
        key="canvasBackGround"
        onClick={[Function]}
      >
        canvas
      </ForwardRef(Button)>
      <ForwardRef(Button)
        color="inherit"
        disabled={false}
        id="cell-color-change"
        key="liveCell"
        onClick={[Function]}
      >
        cell
      </ForwardRef(Button)>
      <ForwardRef(Button)
        color="inherit"
        disabled={false}
        id="grid-color-change"
        key="grid"
        onClick={[Function]}
      >
        grid
      </ForwardRef(Button)>
      <ForwardRef(IconButton)
        aria-label="reset"
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(CancelOutlinedIcon))
          style={
            Object {
              "fontSize": "2rem",
            }
          }
        />
      </ForwardRef(IconButton)>
    </ForwardRef(ButtonGroup)>
  </ForwardRef(DialogActions)>
</ForwardRef(Dialog)>
`;
